### ISSUE TITLE
Provide typed exceptions (java bindings)

### ISSUE THREAD

**Issue by StephenOTT** [digital-asset/daml#14328]
### What is the problem you want to solve?

A lot of extra code is written to wrap the StatusRuntimeException with named/typed exceptions for the ledger api call response.

### What is the solution you would propose?

Provide typed exceptions that provide easy understanding of the business error occurring

Example:

1. Create a User but the user already exists: throw a `UserExistsException` with cause: StatusRuntimeException.
2. Get User but user does not exist: throw a `UserNotFoundException` 
3. Get Contract but does not exist: throw `ContractNotFoundException`

etc etc

Yes we can inspect the status codes and contents of the messages being returned by the ledger api, but it is boiler plate code that needs to be re-written each time.  If you provided typed business exceptions, the boiler plate error handling code would be significantly reduced.

**Comment by S11001001** [comment]
I think this would make sense if we made them checked exceptions, or part of the return type _a la_ `Either`.

As long as we do not, they are _classes_ rather than types (in the Bob Harper sense), and the value seems too limited to me.

NB for @stefanobaghino-da : it is also a significant breaking change.

**Comment by StephenOTT** [comment]
Re breaking chnage: provide a util converter that converts the statusRuntimeException into the proper ~business exception

**Comment by stefanobaghino-da** [comment]
> I think this would make sense if we made them checked exceptions, or part of the return type _a la_ `Either`.
> 
> As long as we do not, they are _classes_ rather than types (in [the Bob Harper sense](https://existentialtype.wordpress.com/2011/03/19/dynamic-languages-are-static-languages/)), and the value seems too limited to me.
> 
> NB for @stefanobaghino-da : it is also a significant breaking change.

This would be a significant breaking change if we made the exceptions checked, right? If we didn't and we made our specific exceptions subtypes of `StatusRuntimeException` there would be not breakage, if I understand correctly. The value I would see here would be to enrich the information provided by gRPC (https://grpc.github.io/grpc-java/javadoc/io/grpc/StatusRuntimeException.html) with our own error-specific ones, allowing users to access fields that would otherwise need to be interpreted [based on the documentation](https://docs.daml.com/canton/usermanual/error_codes.html), like the correlation ID.

**Comment by S11001001** [comment]
> This would be a significant breaking change if we made the exceptions checked, right?

Yes. We have untyped errors right now; if we typed those errors, that would be a breaking change.

> The value I would see here would be to enrich the information provided by gRPC

That can be done with utility functions; we don't need classes for that.

I don't mind the breaking change so much; I bring it up because I know it is generally a concern. But defining a bunch of types, and then not even using those _types_ anywhere, has a real "digging holes and filling them back in" sort of feel to it.

**Comment by daravep** [comment]
@StephenOTT Please have a read about our error codes: https://docs.daml.com/app-dev/grpc/error-codes.html#work-with-error-codes

I would strongly advise to not invent another way to handle errors, but rather make the error-code information accessible. 

1. Every error comes with an error category: https://docs.daml.com/app-dev/grpc/error-codes.html#transientserverfailure 
2. Every error category has an explicit advise on how to deal with that error. Example: https://docs.daml.com/app-dev/grpc/error-codes.html#invalidgivencurrentsystemstateresourceexists
3. Ideally, we surface that retry information.
4. We also need to include the rich error information. E.g. if you get contention on keys, the rich GRPC error contains the information on which key created the contention.
5. Here is our crappy scala code to decode the error-codes into a usable form: https://github.com/DACH-NY/canton/blob/main/community/common/src/main/scala/com/digitalasset/canton/error/DecodedRpcStatus.scala#L22

**Comment by daravep** [comment]
One additional note: we might want to add something like 
```
class ChaosInterceptor
    extends ServerInterceptor {

  override def interceptCall[ReqT, RespT](
      call: ServerCall[ReqT, RespT],
      headers: Metadata,
      next: ServerCallHandler[ReqT, RespT],
  ): ServerCall.Listener[ReqT] = {
     // inject random outages
  }
```
we should add that and turn it on by default on the sandbox. if you build applications against it, it will inject random failures to force you to think about error handling (and let you test your error handling).
